lib/src/diff.rs

```diff
@@ -42,21 +45,22 @@ pub fn find_line_ranges(text: &[u8]) -> Vec<Range<usize>> {
     ranges
 }
 
-fn is_word_byte(b: u8) -> bool {
+fn is_word_grapheme(g: &str) -> bool {
     // TODO: Make this configurable (probably higher up in the call stack)
-    matches!(b, b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_')
+    matches!(g.as_bytes()[0], b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_')
+
 }
 
 pub fn find_word_ranges(text: &[u8]) -> Vec<Range<usize>> {
     let mut word_ranges = vec![];
     let mut word_start_pos = 0;
     let mut in_word = false;
-    for (i, b) in text.iter().enumerate() {
-        if in_word && !is_word_byte(*b) {
+    for (_, (i, g)) in String::from_utf8_lossy(text).grapheme_indices(true).enumerate() {
+        if in_word && !is_word_grapheme(g) {
```

Comment:
`from_utf8_lossy()` replaces invalid characters, so indices of the following bytes will be shifted. We'll probably need to use `str::from_utf8()`, and skip invalid bytes.

Alternatively, we can assume 0x80-0xff as a word. That's simpler and should work well for source code. 
https://www.mercurial-scm.org/repo/hg/file/6.3.3/mercurial/patch.py#l51

Reply:
Ah, really good point. So, IIUC, you're proposing that we either
1. Switch to using `from_utf8` which entails handling and propagating the `Utf8Error` around as necessary; or
2. Strip out the grapheme stuff and just treat `0x80`–`0xff` as word-characters?

The second option is a really cute workaround to handle multi-byte utf8 chars, which I kind of like. That said, handling grapheme clusters is somewhat unavoidable in the long run, so I kind of learn towards biting the bullet and properly handling invalid utf8 errors. Thoughts?

If we go with option 1, there's also the question of semantics around how we want to handle `Utf8Error`.

Reply:
> you're proposing that we either
> 
>     1. Switch to using `from_utf8` which entails handling and propagating the `Utf8Error` around as necessary; or
> 
>     2. Strip out the grapheme stuff and just treat `0x80`–`0xff` as word-characters?

Yes. For the option 1, `find_word_ranges()` will need to handle `Utf8Error`, and redo utf8 decoding up to `valid_up_to()`. The `error_len()` bytes from `valid_up_to()` will be added as a word range, then start utf8 decoding from there.

> The second option is a really cute workaround to handle multi-byte utf8 chars, which I kind of like. That said, handling grapheme clusters is somewhat unavoidable in the long run, so I kind of learn towards biting the bullet and properly handling invalid utf8 errors.

I have no idea if we need a unicode-aware word splitting. Maybe it's useful for Latin languages, but I'm not familiar with such languages. For me, the option 2 is good enough.

Reply:
Option 2 works just fine for now. My main concern is the comment on `is_word_byte`:

```
    // TODO: Make this configurable (probably higher up in the call stack)
```

which perhaps involves design decisions around Unicode-aware segmentation anyway. For now I'll just go with your Option 2!